File: lc_go/lcGo/DFS/findCircleNum.go
package main

type Union struct {
	parent []int
	count  int
}

func BuildUnionFind(n int) *Union {
	parent := make([]int, n)
	for i := range parent {
		parent[i] = i
	}
	return &Union{parent: parent, count: n}
}

func (u *Union) unionSet(x, y int) {
	rootX := u.find(x)
	rootY := u.find(y)

	if rootX != rootY {
		u.parent[rootX] = rootY
		u.count--
	}
}

func (u *Union) getCount() int {
	return u.count
}

func (u *Union) find(x int) int {
	if u.parent[x] != x {
		u.parent[x] = u.find(u.parent[x])
	}
	return u.parent[x]
}

func findCircleNum(isConnected [][]int) int {
	n := len(isConnected)

	u := BuildUnionFind(n)

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if isConnected[i][j] == 1 {
				u.unionSet(i, j)
			}
		}
	}
	return u.getCount()
}

func main() {

}

File: lc_cpp/lcCpp/DFS/findCircleNum.cpp
//
// Created by 24276 on 2025/8/7.
//
#include <vector>

class Union {
private:
    std::vector<int> parent;
    int count;
public:
    Union(int n) {
        parent.resize(n);
        count = n;
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
    }
    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }
    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if( rootX != rootY) {
            parent[rootX] = rootY;
            --count;
        }
    }
    int getCount() const {
        return count;
    }

};
class find_circle_num {
public:
    int findCircleNum (std::vector<std::vector<int>>& isConnected) {
        int n = isConnected.size();
        Union u(n);
        for (int i = 0; i < n;++i) {
            for(int j = i + 1; j < n;j++) {
                if (isConnected[i][j] == 1) {
                    u.unionSet(i, j);
                }
            }
        }
        return u.getCount();
    }
};

int main() {
    return 0;
}
